ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/Update.hs

```diff
@@ -114,15 +114,22 @@ toRealPoint (Weaken ap)      = toRealPoint ap
 -- We take in the entire 'LedgerDB' because we record that as part of errors.
 applyBlock :: forall m c l blk. (ApplyBlock l blk, Monad m, c)
            => LedgerCfg l
+           -> LedgerEventHandler m l
            -> Ap m l blk c
            -> LedgerDB l -> m l
-applyBlock cfg ap db = case ap of
+applyBlock cfg eventHandler@LedgerEventHandler{handleLedgerEvent} ap db = case ap of
     ReapplyVal b ->
       return $
         tickThenReapply cfg b l
-    ApplyVal b ->
-      either (throwLedgerError db (blockRealPoint b)) return $ runExcept $
-        tickThenApply cfg b l
+    ApplyVal b -> do
+      result <- either (throwLedgerError db (blockRealPoint b)) return $ runExcept $
+        tickThenApplyLedgerResult cfg b l
+      forM_ (lrEvents result) $
```

Comment:
Am I correct that this is the only place you're actually invoking the given event handler?

Reply:
Correct. Though not all call to `applyBlock` will have an handler. In some situations, like when blocks are re-applied after a restart or, when blocks are applied to our own local chain, we pass down `discardEvent` to avoid yielding unnecessary events. 

And we'd love your insights on that. The idea is to only yield event once for each block - modulo rollbacks.